File: Data Structures & Algorithms/same-binary-tree/submission-0.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn is_same_tree(p: Option<Rc<RefCell<TreeNode>>>, q: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if p.is_none() && q.is_none(){
            return true;
        } 

        if p.is_none() || q.is_none() {
            return false;
        }

        let p = p.unwrap();
        let q = q.unwrap();
        let p_borrow = p.borrow();
        let q_borrow = q.borrow();

        if p_borrow.val != q_borrow.val {
            return false
        }

        Self::is_same_tree(p_borrow.right.clone(),q_borrow.right.clone()) && 
        Self::is_same_tree(p_borrow.left.clone(),q_borrow.left.clone())
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool compareTree(TreeNode* root, TreeNode* subRoot) {
        if (subRoot == nullptr && root == nullptr) return true;
        if (subRoot == nullptr || root == nullptr ) return false;
        if (root->val != subRoot->val) return false;
        return compareTree(root->left, subRoot->left) && compareTree(root->right, subRoot->right);
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root == nullptr) return false;
        bool compare = false;
        if (root->val == subRoot->val) {
            compare = compareTree(root,subRoot);
        }
        return compare || isSubtree(root->left,subRoot) || isSubtree(root->right,subRoot);
    }
};
